lib/qemu_sandbox.ml

```diff
@@ -0,0 +1,221 @@
+  let rec aux () =
+    Lwt_unix.read src buf 0 (Bytes.length buf) >>= function
+    | 0 -> Lwt.return_unit
+    | n -> Build_log.write dst (Bytes.sub_string buf 0 n) >>= aux
+  in
+  aux ()
+
+type guest_os =
+  | Linux
+  | OpenBSD
+  | Windows
+[@@deriving sexp]
+
+type guest_arch =
+  | Amd64
+  | Riscv64
+[@@deriving sexp]
+
+type t = {
+  qemu_cpus : int;
+  qemu_memory : int;
+  qemu_guest_os : guest_os;
+  qemu_guest_arch : guest_arch;
+  qemu_boot_time : int;
+}
+
+type config = {
+  cpus : int;
+  memory : int;
+  guest_os : guest_os;
+  guest_arch : guest_arch;
+  boot_time : int;
+} [@@deriving sexp]
+
+let get_free_port () =
+  let fd = Unix.socket PF_INET SOCK_STREAM 0 in
+  let () = Unix.bind fd (ADDR_INET(Unix.inet_addr_loopback, 0)) in
+  let sa = Unix.getsockname fd in
+  let () = Unix.close fd in
+  match sa with
+  | ADDR_INET (_, n) -> string_of_int n
+  | ADDR_UNIX _ -> assert false;;
+
+let run ~cancelled ?stdin ~log t config result_tmp =
+  let pp f = Os.pp_cmd f ("", config.Config.argv) in
+
+  let extra_mounts = List.map (fun { Config.Mount.src; _ } ->
+    ["-drive"; "file=" ^ src / "rootfs" / "image.qcow2" ^ ",if=virtio"]
+  ) config.Config.mounts |> List.flatten in
+
+  Os.with_pipe_to_child @@ fun ~r:qemu_r ~w:qemu_w ->
+  let qemu_stdin = `FD_move_safely qemu_r in
+  let qemu_monitor = Lwt_io.(of_fd ~mode:output) qemu_w in
+  let port = get_free_port () in
+  let qemu_binary = match t.qemu_guest_arch with
+    | Amd64 -> [ "qemu-system-x86_64"; "-machine"; "accel=kvm,type=pc"; "-cpu"; "host"; "-display"; "none";
+                 "-device"; "virtio-net,netdev=net0" ]
+    | Riscv64 -> [ "qemu-system-riscv64"; "-machine"; "type=virt"; "-nographic";
+                   "-bios"; "/usr/lib/riscv64-linux-gnu/opensbi/generic/fw_jump.bin";
+                   "-kernel"; "/usr/lib/u-boot/qemu-riscv64_smode/uboot.elf";
+                   "-device"; "virtio-net-device,netdev=net0";
+                   "-serial"; "none"] in
+  let network = match config.network with
+    | [ "host" ] -> ""
+    | _ -> "restrict=yes," in
+  let cmd = qemu_binary @ [
+              "-monitor"; "stdio";
+              "-m"; (string_of_int t.qemu_memory) ^ "G";
+              "-smp"; string_of_int t.qemu_cpus;
+              "-netdev"; "user,id=net0," ^ network ^ "hostfwd=tcp::" ^ port ^ "-:22";
+              "-drive"; "file=" ^ result_tmp / "rootfs" / "image.qcow2" ^ ",if=virtio" ]
+              @ extra_mounts in
+  let _, proc = Os.open_process ~stdin:qemu_stdin ~stdout:`Dev_null ~pp cmd in
+
+  let ssh = ["ssh"; "opam@localhost"; "-p"; port; "-o"; "NoHostAuthenticationForLocalhost=yes"] in
+
+  let rec loop = function
+    | 0 -> Lwt_result.fail (`Msg "No connection")
+    | n ->
+      Os.exec_result ~pp (ssh @ ["exit"]) >>= function
+      | Ok _ -> Lwt_result.ok (Lwt.return ())
+      | _ -> Lwt_unix.sleep 1. >>= fun _ -> loop (n - 1) in
+  Lwt_unix.sleep 5. >>= fun _ ->
+  loop t.qemu_boot_time >>= fun _ ->
```

Comment:
I don't think it matters.  A failure here would indicate that `qemu_boot_time` is too low or the base image is bad.  The next `ssh` commands would also fail and we would then orderly clean up the QEMU machine with the `quit` command.

Reply:
I generally have a preference for failing early, as it helps with debugging and can sometimes make it easier to improve fault tolerance later on. But I am convinced it won't cause problems here, so won't press the point.

A comment here to explain that reasoning may help future readers, but I don't insist on that either in case you think it's not needed.